### ISSUE TITLE
Discuss if MeshPy should move to a src layout

### ISSUE THREAD

**Issue by davidrudlstorfer** [imcs-compsim/meshpy#159]
Currently MeshPy is structured in a flat layout, i.e. we have the meshpy module at the top as well as "external"

https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/

https://www.reddit.com/r/learnpython/comments/wpkr4q/using_a_src_directory_for_a_python_package/

https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-structure.html

For simplification of the entire project and its structure I propose to switch to a src based layout

```
meshpy/
├─ src/
│  ├─ pybind11 (as git submodule)/
│  ├─ meshpy/
│  │  ├─ core/
│  │  │  ├─ inputfile.py
│  │  │  ├─ function.py
│  │  │  ├─ ...
│  │  ├─ four_c/
│  │  ├─ geometric_search/
│  │  ├─ .../
│  ├─ utilities/

```

The structure within the `tests/` folder should identically reflect the src structure of meshpy

This would allow us to easily define all modules and also export meshpy as a package

**Comment by davidrudlstorfer** [comment]
Thanks for already looking into this @isteinbrecher! This was the last point on my personal ToDo list for MeshPy which was not yet turned into an issue here on github 😅

For sure this should only happen once the testing is ported 🙏🏻

**Comment by davidrudlstorfer** [comment]
One question or rather discussion because this would be a breaking change for MeshPy

Is there actually a valid reason why we specify so many subpackages within MeshPy (i.e., abaqus, four_c, geometric_search)?

Yesterday I setup a local branch regarding the API documentation for MeshPy - once again I ran into the problem that MeshPy is not a single package but multiple packages - this is not really supported by any documentation. This will also lead to problems down the road if we want to publish the package to PyPi as far as i know

In my opinion we should just have one package, i.e. MeshPy. We can still access all the stuff via, for example, `meshpy.rotation.blablub`

So is there actually a reason behind all of that or can this also be simplified when moving to the src layout?

I would strongly add the move here, because once we change the layout structure of MeshPy - other 3rd party tools need to adjust imports - this could then be done in one sweep

Curious about your opinion @isteinbrecher

**Comment by isteinbrecher** [comment]
Good point. Short answer, there is no reason to have sub packages. I like the "structuring" that comes with accessing things via `meshpy.sub_namespace.stuff` and I think the sub packages were the first thing I tried (a long time ago) and they worked. So I am totally fine with resolving this.

**Comment by davidrudlstorfer** [comment]
Thanks for the fast reply - then I can further look into this during the restructuring.

**Comment by davidrudlstorfer** [comment]
This one is completed with #238 - I will open a follow up issue with the points raised in #238
